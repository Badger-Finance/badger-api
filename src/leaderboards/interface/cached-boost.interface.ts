import { attribute, hashKey, rangeKey, table } from '@aws/dynamodb-data-mapper-annotations';
import { LEADERBOARD_DATA } from '../../config/constants';

@table(LEADERBOARD_DATA)
export class CachedBoost {
  @hashKey()
  leaderboard!: string;

  @rangeKey()
  rank!: number;

  @attribute()
  boost!: number;

  @attribute()
  stakeRatio!: number;

  @attribute()
  address!: string;
}
