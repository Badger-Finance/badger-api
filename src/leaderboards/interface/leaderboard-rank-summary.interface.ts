import { attribute } from "@aws/dynamodb-data-mapper-annotations";

export class LeaderboardRankSummary {
  @attribute()
  badgerType!: string;

  @attribute()
  amount!: number;
}
