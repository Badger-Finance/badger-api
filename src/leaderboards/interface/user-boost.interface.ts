export interface UserBoost {
  address: string;
  boost: number;
  stakeRatio: number;
  nftMultiplier: number;
  nativeBalance: number;
  nonNativeBalance: number;
}
