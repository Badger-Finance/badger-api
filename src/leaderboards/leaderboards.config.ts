import { BadgerType } from './enums/badger-type.enum';
import { BadgerTypeMap } from './types/badger-type-map';

export const BADGER_RANKS: BadgerTypeMap = {
  [BadgerType.Basic]: 1,
  [BadgerType.Neo]: 20,
  [BadgerType.Hero]: 200,
  [BadgerType.Hyper]: 600,
  [BadgerType.Frenzy]: 1400,
};

export function getBadgerType(score: number): BadgerType {
  if (score >= BADGER_RANKS[BadgerType.Frenzy]) {
    return BadgerType.Frenzy;
  }
  if (score >= BADGER_RANKS[BadgerType.Hyper]) {
    return BadgerType.Hyper;
  }
  if (score >= BADGER_RANKS[BadgerType.Hero]) {
    return BadgerType.Hero;
  }
  if (score >= BADGER_RANKS[BadgerType.Neo]) {
    return BadgerType.Neo;
  }
  return BadgerType.Basic;
}
