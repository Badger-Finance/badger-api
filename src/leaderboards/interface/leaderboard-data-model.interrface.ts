import { Description, Example, Property, Title } from '@tsed/schema';
import { CachedBoost } from './cached-boost.interface';
import { LeaderBoardData } from './leaderboard-data.interrface';

export class LeaderBoardDataModel implements LeaderBoardData {
  @Title('data')
  @Description('Leaderboard page data')
  @Example([
    {
      leaderboard: 'bosst',
      rank: 1,
      address: '0xdeadbeef',
      boost: 3,
      stakeRatio: 100,
      nftMultiplier: 3,
      nativeBalance: 300234.23,
      nonNativeBalance: 3245.12,
    },
  ])
  @Property()
  public data: CachedBoost[];

  @Title('page')
  @Description('Leaderboard page')
  @Example(1)
  @Property()
  public page: number;

  @Title('size')
  @Description('Leaderboard page size')
  @Example(20)
  @Property()
  public size: number;

  @Title('count')
  @Description('Leaderboard entries')
  @Example(3087)
  @Property()
  public count: number;

  @Title('maxPage')
  @Description('Maximum leaderboard page')
  @Example(65)
  @Property()
  public maxPage: number;

  constructor(data: LeaderBoardData) {
    this.data = data.data;
    this.page = data.page;
    this.size = data.size;
    this.count = data.count;
    this.maxPage = data.maxPage;
  }
}
