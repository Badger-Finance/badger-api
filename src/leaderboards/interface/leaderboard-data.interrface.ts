import { CachedBoost } from './cached-boost.interface';

export interface LeaderBoardData {
  data: CachedBoost[];
  page: number;
  size: number;
  count: number;
  maxPage: number;
}
