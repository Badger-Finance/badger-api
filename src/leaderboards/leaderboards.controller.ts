import { LeaderboardSummary, Network } from "@badger-dao/sdk";
import { Controller, Get, Inject, QueryParams } from "@tsed/common";
import { ContentType, Hidden } from "@tsed/schema";

import { Chain } from "../chains/config/chain.config";
import { LeaderBoardsService } from "./leaderboards.service";

@Controller("/leaderboards")
export class LeaderBoardsController {
  @Inject()
  leaderBoardsService!: LeaderBoardsService;

  @Get("")
  @Hidden()
  @ContentType("json")
  async getLeaderBoardSummary(@QueryParams("chain") chain?: Network): Promise<LeaderboardSummary> {
    return this.leaderBoardsService.fetchLeaderboardSummary(Chain.getChain(chain));
  }
}
