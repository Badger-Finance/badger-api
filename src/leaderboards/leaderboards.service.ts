import { BadgerTypeMap, LeaderboardSummary } from '@badger-dao/sdk';
import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { queryLeaderboardSummary } from './leaderboards.utils';

@Service()
export class LeaderBoardsService {
  async fetchLeaderboardSummary(chain: Chain): Promise<LeaderboardSummary> {
    const cachedSummary = await queryLeaderboardSummary(chain);
    const summary = Object.fromEntries(
      cachedSummary.rankSummaries.map((s) => [s.badgerType, s.amount])
    ) as BadgerTypeMap;
    return {
      summary,
      updatedAt: cachedSummary.updatedAt
    };
  }
}
