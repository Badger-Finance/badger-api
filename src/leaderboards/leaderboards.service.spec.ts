import { PlatformTest } from '@tsed/common';
import { setupMapper, TEST_ADDR } from '../test/tests.utils';
import { LeaderBoardType } from './enums/leaderboard-type.enum';
import { CachedBoost } from './interface/cached-boost.interface';
import { LeaderBoardData } from './interface/leaderboard-data.interrface';
import { LeaderBoardsService } from './leaderboards.service';

describe('leaderboards.service', () => {
  let service: LeaderBoardsService;

  beforeAll(async () => {
    await PlatformTest.create();
    service = PlatformTest.get<LeaderBoardsService>(LeaderBoardsService);
  });

  afterEach(PlatformTest.reset);

  function randomCachedBoosts(count: number): CachedBoost[] {
    const boosts = [];
    for (let i = 0; i < count; i += 1) {
      boosts.push(Object.assign(new CachedBoost(), {
        leaderboard: LeaderBoardType.BadgerBoost,
        rank: i,
        address: TEST_ADDR,
        boost: 2000 - (i * 10),
        nftMultiplier: 1,
        stakeRatio: 1 - (i * 0.01),
        nativeBalance: 100000 / (i + 1),
        nonNativeBalance: 250000 / (i + 1),
      }));
    }
    return boosts;
  }

  function expectedData(expected: CachedBoost[], page?: number, size?: number): LeaderBoardData {
    const pageNumber = page || 0;
    const pageSize = size || 20;
    const offset = pageNumber > 0 ? 1 : 0;
    const start = pageNumber * pageSize + offset;
    const end = start + pageSize - offset;
    const maxPage = parseInt((expected.length / pageSize).toString());
    const data = expected.slice(start, end);
    return {
      data,
      page: pageNumber,
      size: pageSize,
      count: data.length,
      maxPage,
    }
  }

  describe('loadFullLeaderBoard', () => {
    describe('no saved leaderboard entries', () => {
      it('returns an empty array', async () => {
        setupMapper([]);
        const result = await service.loadFullLeaderBoard();
        expect(result).toMatchObject([]);
      });
    });
    describe('saved leaderboard entries', () => {
      it('returns all saved leaderboard entries', async () => {
        const expected = randomCachedBoosts(50);
        setupMapper(expected);
        const result = await service.loadFullLeaderBoard();
        expect(result).toMatchObject(expected);
      });
    });
  });

  describe('loadLeaderboardEntries', () => {
    describe('no saved leaderboard entries', () => {
      it('returns an empty array', async () => {
        setupMapper([]);
        const result = await service.loadLeaderboardEntries();
        expect(result).toMatchObject(expectedData([]));
      });
    });
    describe('saved 50 leaderboard entries', () => {
      const saved = randomCachedBoosts(50);
      setupMapper(saved);
      describe('no parameters', () => {
        it('returns an array of first 20 boosts', async () => {
          const result = await service.loadLeaderboardEntries();
          expect(result).toMatchObject(saved.slice(20));
        });
      });
    });
  });
});
