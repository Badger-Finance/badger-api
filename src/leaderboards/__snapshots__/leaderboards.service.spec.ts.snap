// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaderboards.service fetchLeaderboardSummary returns the current leaderboard summary for the requested chain 1`] = `
Object {
  "summary": Object {
    "basic": 1000,
    "frenzy": 40,
    "hero": 35,
    "hyper": 25,
    "neo": 20,
  },
  "updatedAt": 133742069,
}
`;
